anospp_analysis/nn.py

```diff
@@ -0,0 +1,501 @@
+
+    ref_hap_df, af_c, af_i, af_f, true_multi_targets, \
+        colors_coarse, colors_int, colors_fine = prep_reference_index(\
+        args.reference, path_to_refversion=args.path_to_refversion)
+
+    kmers = construct_unique_kmer_table(mosq_hap_df, int(args.kmer_length))
+    ref_kmers = construct_unique_kmer_table(ref_hap_df, int(args.kmer_length))
+
+    error_seqs = identify_error_seqs(mosq_hap_df, kmers, int(args.kmer_length), int(args.n_error_snps))
+    non_error_hap_df = mosq_hap_df[~mosq_hap_df.seqid.isin(error_seqs)]
+    non_error_hap_df = recompute_haplotype_coverage(non_error_hap_df)
+    non_error_hap_df.to_csv(f'{args.outdir}/non_error_haplotypes.tsv', index=False, sep='\t')
+    
+    nndict = find_nn_unique_haps(non_error_hap_df, kmers, ref_hap_df, ref_kmers)
+    nn_df = pd.DataFrame.from_dict(nndict, orient='index', columns=['nn_id_array', 'nn_dist'])
+    nn_df['nn_id'] = ['|'.join(map(str, l)) for l in nn_df.nn_id_array]
+    nn_df[['nn_id', 'nn_dist']].to_csv(f'{args.outdir}/nn_dictionary.tsv', sep='\t')
+
+    result_coarse, result_int, result_fine, test_samples = perform_nn_assignment_samples(\
+        non_error_hap_df, ref_hap_df, nndict, af_c, af_i, af_f, args.allelism_normalisation)
+    result_coarse.to_csv(f"{args.outdir}/assignment_coarse.tsv", sep='\t')
+    result_int.to_csv(f"{args.outdir}/assignment_intermediate.tsv", sep='\t')
+    result_fine.to_csv(f"{args.outdir}/assignment_fine.tsv", sep='\t')
+
+    comb_stats_df = generate_hard_calls(comb_stats_df, non_error_hap_df, test_samples, \
+        result_coarse, result_int, result_fine, true_multi_targets, args.nn_assignment_threshold, \
+            args.medium_contamination_read_count_threshold, \
+            args.medium_contamination_multi_allelic_threshold, \
+            args.high_contamination_multi_allelic_threshold)
+
+    logging.info(f'writing assignment results to {args.outdir}')
+    comb_stats_df.to_csv(f'{args.outdir}/nn_assignment.tsv', index=False, sep='\t')
+
+    if not bool(args.no_plotting):
+        for result, level, colors in zip([result_coarse, result_int, result_fine], \
+                             ['coarse', 'intermediate', 'fine'], [colors_coarse, \
+                                                colors_int, colors_fine]):
+            fig, _ = plot_assignment_proportions(result, level, colors, args.nn_assignment_threshold)
+            fig.savefig(f'{args.outdir}/{level}_assignment.png')
+
+    logging.info('All done!')
+
+    
+def main():
+    
+    parser = argparse.ArgumentParser("NN assignment for ANOSPP sequencing data")
+    parser.add_argument('-a', '--haplotypes', help='Haplotypes tsv file', required=True)
+    parser.add_argument('-m', '--manifest', help='Samples manifest tsv file', required=True)
+    parser.add_argument('-s', '--stats', help='DADA2 stats tsv file', required=True)
+    parser.add_argument('-r', '--reference', help='Reference index version - currently a directory name.\
```

Comment:
If I understood the use correctly, -r and -p are only used together to form path to reference index, so it seems reasonable to combine them into a single parameter

It is also reasonable to record the version of the reference somewhere in the results - I would prefer to have some summary file with key stats (reference version, samples total, samples in assignment, number of lineages assigned at each level, number of unassigned/rainbow samples etc), but adding it as a column to nn_assignment.tsv is an option too.

In order to achieve that, it seems most reasonable to include version info in the reference dataset, e.g. as a small version.txt file. This way, we won't depend on directory names

Reply:
regarding `-r` and `-p`, for now yes. But eventually I want to read the reference version from a publicly accessible storage so in practice I expect that most people virtually always use the default path pointing to the storage location, but the `-p` function would allow me to run some testing on updated reference databases before pushing them to the public storage. 

I like the idea of having a small summary file as output - I'll draft one.

Reply:
I could remove the `-p` option to specify the path and only use `--path_to_refversion`; that way it's maybe clearer that users are generally not expected to change it, just like all the thresholds.